core/execution/liquidation/engine.go

```diff
@@ -0,0 +1,307 @@
+	if len(closed) == 0 {
+		return nil, nil, nil
+	}
+	// netork is most likely going to hold an actual position now, let's set up the time step when we attempt to dispose
+	// of (some) of the volume
+	if e.pos.open == 0 || e.nextStep.IsZero() {
+		e.nextStep = e.tSvc.GetTimeNow().Add(e.cfg.DisposalTimeStep)
+	}
+	mps := make([]events.MarketPosition, 0, len(closed))
+	parties := make([]string, 0, len(closed))
+	// order events here
+	orders := make([]events.Event, 0, len(closed)*2)
+	// trade events here
+	trades := make([]events.Event, 0, len(closed))
+	netTrades := make([]*types.Trade, 0, len(closed))
+	now := e.tSvc.GetTimeNow()
+	for _, cp := range closed {
+		e.pos.open += cp.Size()
+		// get the orders and trades so we can send events to update the datanode
+		o1, o2, t := e.getOrdersAndTrade(cp, idgen, now, mp, mmp)
+		orders = append(orders, events.NewOrderEvent(ctx, o1), events.NewOrderEvent(ctx, o2))
+		trades = append(trades, events.NewTradeEvent(ctx, *t))
+		netTrades = append(netTrades, t)
+		// add the confiscated balance to the fee pool that can be taken from the insurance pool to pay fees to
+		// the good parties when the network closes itself out.
+		mps = append(mps, cp)
+		parties = append(parties, cp.Party())
+	}
+	// send order events
+	e.broker.SendBatch(orders)
+	// send trade events
+	e.broker.SendBatch(trades)
+	// the network has no (more) remaining open position -> no need for the e.nextStep to be set
+	if e.pos.open == 0 {
+		e.nextStep = time.Time{}
+	}
+	return mps, parties, netTrades
+}
+
+func (e *Engine) UpdateMarkPrice(mp *num.Uint) {
+	e.pos.price = mp
+}
+
+func (e *Engine) GetNetworkPosition() events.MarketPosition {
+	return e.pos
+}
+
+func (e *Engine) UpdateNetworkPosition(trades []*types.Trade) {
+	sign := int64(1)
+	if e.pos.open < 0 {
+		sign *= -1
+	}
+	for _, t := range trades {
+		delta := int64(t.Size) * sign
+		e.pos.open -= delta
+	}
+	if e.pos.open == 0 {
+		e.nextStep = time.Time{}
+	} else if e.nextStep.IsZero() {
+		e.nextStep = e.tSvc.GetTimeNow().Add(e.cfg.DisposalTimeStep)
+	}
+}
+
+func (e *Engine) getOrdersAndTrade(pos events.Margin, idgen IDGen, now time.Time, price, dpPrice *num.Uint) (*types.Order, *types.Order, *types.Trade) {
+	tSide, nSide := types.SideSell, types.SideBuy // one of them will have to sell
+	s := pos.Size()
+	size := uint64(s)
+	if s < 0 {
+		size = uint64(-s)
+		// swap sides
+		nSide, tSide = tSide, nSide
+	}
+	var buyID, sellID, buyParty, sellParty string
+	order := types.Order{
+		ID:            idgen.NextID(),
+		MarketID:      e.mID,
+		Status:        types.OrderStatusFilled,
+		Party:         types.NetworkParty,
+		Price:         price,
+		OriginalPrice: dpPrice,
+		CreatedAt:     now.UnixNano(),
+		Reference:     "close-out distressed",
+		TimeInForce:   types.OrderTimeInForceFOK, // this is an all-or-nothing order, so TIME_IN_FORCE == FOK
```

Comment:
Isn't the network order supposed to be IOC now?

Reply:
eh I'm stupid that's the order with the distressed party.

Reply:
The order the network places to close its own position, yes, and that's an IOC order. These are the "fake" orders that populate the order/trade events for the closeout of the distressed parties.